cumulusci/tasks/push/tests/test_push_tasks.py

```diff
@@ -0,0 +1,481 @@
+    )
+
+
+@pytest.fixture
+def package_push_request_cancel():
+    return PackagePushRequest(
+        push_api=mock.MagicMock(),
+        version="1.2.3",
+        start_time="12:03",
+        status="Canceled",
+        sf_id=SF_ID,
+    )
+
+
+def test_parse_version():
+    task = create_task(BaseSalesforcePushTask, options={})
+    assert task._parse_version("1.2") == {
+        "major": "1",
+        "minor": "2",
+        "patch": None,
+        "build": None,
+        "state": "Released",
+    }
+    assert task._parse_version("1.2,Beta 3") == {
+        "major": "1",
+        "minor": "2",
+        "patch": None,
+        "build": "Beta 3",
+        "state": "Beta",
+    }
+    assert task._parse_version("1.2,(Beta 3)") == {
+        "major": "1",
+        "minor": "2",
+        "patch": None,
+        "build": "(Beta 3",
+        "state": "Beta",
+    }
+    assert task._parse_version("1.2") == {
+        "major": "1",
+        "minor": "2",
+        "patch": None,
+        "build": None,
+        "state": "Released",
+    }
+
+
+def test_get_version():
+    task = create_task(BaseSalesforcePushTask, options={})
+    assert task._get_version(
+        MetadataPackage(
+            push_api=mock.MagicMock(), name=NAME, sf_id=SF_ID, namespace=NAMESPACE
+        ),
+        "1.2.3.4",
+    )
+    assert task._get_version(
+        MetadataPackage(
+            push_api=mock.MagicMock(), name=NAME, sf_id=SF_ID, namespace=NAMESPACE
+        ),
+        "1.2,Beta 3",
+    )
+
+
+def test_get_version_error():
+    package = mock.MagicMock()
+    package.get_package_version_objs.return_value = None
+    task = create_task(BaseSalesforcePushTask, options={})
+    with pytest.raises(PushApiObjectNotFound):
+        task._get_version(package, VERSION)
+
+
+def test_get_package():
+    task = create_task(BaseSalesforcePushTask, options={})
+    task.push = mock.MagicMock()
+    task.push.get_package_objs.return_value = [
+        {"push_api": "123", "sf_id": SF_ID, "name": NAME, "namespace": NAMESPACE}
+    ]
+    assert task._get_package(NAMESPACE) == {
+        "push_api": "123",
+        "sf_id": SF_ID,
+        "name": NAME,
+        "namespace": NAMESPACE,
+    }
+
+
+def test_get_package_error():
+    task = create_task(BaseSalesforcePushTask, options={})
+    task.push = mock.MagicMock()
+    task.push.get_package_objs.return_value = None
+    with pytest.raises(PushApiObjectNotFound):
+        task._get_package(NAMESPACE)
+
+
+def test_schedule_push_org_list_get_orgs():
+    with open(ORG_FILE, "w") as file:
+        file.write(f"{ORG_FILE_TEXT}")
+    task = create_task(
+        SchedulePushOrgList,
+        options={
+            "orgs": ORG_FILE,
+            "version": VERSION,
+            "namespace": NAMESPACE,
+            "start_time": datetime.datetime.now(),
+            "batch_size": 10,
+        },
+    )
+    assert task._get_orgs() == ["00DS0000003TJJ6MAO", "00DS0000003TJJ6MAL"]
+    os.remove(ORG_FILE)
```

Comment:
This is a non-guaranteed way to do cleanup in a test, because if there's an error or assertion failure in the test, `os.remove` will not run.

You've got a few options for how to make it better:
1. Wrap the entire test in try/finally and put os.remove in the finally block.
2. Wrap the entire test in `with temporary_dir() as path:` and make sure you create the file inside the `path` directory (the temporary_dir context manager takes care of deleting the directory when it exits)
3. Create a separate fixture to handle the file:

```
@pytest.fixture
def org_file():
    with open(ORG_FILE, "w") as file:
        file.write(ORG_FILE_TEXT)
    try:
        yield  # this is where the test using the fixture runs
    finally:
        os.remove(ORG_FILE)
```

I like option 3 because then you don't have to repeat the logic for each test that uses the fixture.

Reply:
pytest provides the [`tmp_path` fixture](https://docs.pytest.org/en/stable/tmpdir.html#the-tmp-path-fixture), which does what you want.

Reply:
@jstvz TIL!